### PR TITLE
StringView: Implement wildcard pattern matching

### PR DESCRIPTION
**Pull request by howar6hill**
Wildcard pattern matching is widely used in many programs. It would be beneficial to provide this functionality.

### CODE REVIEW

**Comment by bugaevc** [comment]
There's a lot of functionality, currently duplicated, that I would like to see shared between `String` and `StringView`. The most C++ way, I think, would be to move them to a `StringBase` class that both would inherit from.

**Comment by awesomekling** [comment]
> There's a lot of functionality, currently duplicated, that I would like to see shared between `String` and `StringView`. The most C++ way, I think, would be to move them to a `StringBase` class that both would inherit from.

Sure, two approaches that might work:
* A CRTP (`StringBase<T>`) base class with common functionality as member functions
* A `StringFunctions` utility namespace with common functionality as free functions

**Comment by bugaevc** [comment]
I don't see what you'd need CRTP here for?

**Comment by awesomekling** [comment]
CRTP would allow the subclasses to decide their own memory layout.

**Comment by bugaevc** [comment]
Wait, does C++ not allow you to have a pure virtual method without having to actually use vtables and dynamic dispatch? So CRTP here would basically be a workaround for that. Makes some sense.

**Comment by bugaevc** [comment]
> Wait, does C++ not allow you to have an abstract method without having to actually use vtables and dynamic dispatch?

Actually it does make some more sense, because the base class code is not additionally monomorphized *unless* you add a template parameter explicitly, so that's exactly what CRTP does.

**Comment by howar6hill** [comment]
I think putting  this functionality in the `StringFunctions`  namespace would be easier. And common functionality in the `StringFunctions`  namespace would be easier to understand, and easier to maintain.

**Comment by bugaevc** [comment]
But `StringFunctions::matches(foo, bar)` looks so much worse than `foo.matches(bar)`...

**Comment by awesomekling** [comment]
> But `StringFunctions::matches(foo, bar)` looks so much worse than `foo.matches(bar)`...

I meant that we could have the functionality in a utility namespace (`AK::StringFunctions`) and then call it from wrappers. Something like this:

```cpp
bool StringView::matches(const StringView& mask, CaseSensitivity case_sensitivity) const
{
    return StringFunctions::matches(*this, mask, case_sensitivity);
}
```

**Comment by howar6hill** [comment]
I think we should provide a free function, and import it to the global namespace. That way you can pass a C string as parameter, and  don't have to construct a StringView or String object first. We could also provide wrappers in the StringView and String classes, if that looks prettier.

**Comment by bugaevc** [comment]
> That way you can pass a C string as parameter, and don't have to construct a StringView or String object first.

Actually I would prefer if you *had* to convert a C string to a native string type first...

**Comment by howar6hill** [comment]
> > That way you can pass a C string as parameter, and don't have to construct a StringView or String object first.
> 
> Actually I would prefer if you _had_ to convert a C string to a native string type first...

The benefit of it? The other way seems more convenient to me.

**Comment by awesomekling** [comment]
We pass `StringView` whenever possible instead of C strings.
The bad thing about C strings is that you must traverse the whole string to find out how long it is. `StringView` does not have this problem.

**Comment by bugaevc** [comment]
It's not that *bad*, but it's nicer when you do the C string -> `StringView` conversion explicitly (and *that's* where you traverse the memory looking for the null byte) and pass around `StringView` after that.

**Comment by howar6hill** [comment]
I mean the free function signature should be like this
```C++
bool matches(const StringView& str, const StringView& mask, CaseSensitivity case_sensitivity)
```
Then you can pass a C string, StringView or String.

**Comment by awesomekling** [comment]
> I mean the free function signature should be like this
> 
> ```c++
> bool matches(const StringView& str, const StringView& mask, CaseSensitivity case_sensitivity)
> ```
> 
> Then you can pass a C string, StringView or String.

Aha yeah, that makes sense :)

**Comment by howar6hill** [comment]
Moved the wildcard-matching implementation to StringUtils.
Now String and StringView share the same functionality.

**Comment by howar6hill** [comment]
@bugaevc @awesomekling What do you think about the design?

File: AK/StringUtils.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace StringUtils {
+
+    bool matches(const StringView& str, const StringView& mask, CaseSensitivity = CaseSensitivity::CaseInsensitive);
+
+}
+
+}
+
+using AK::CaseSensitivity;
+using AK::StringUtils::matches;
```
Comment (awesomekling) [inline AK/StringUtils.h]: Let's not bring this into the global namespace.

**Comment by howar6hill** [comment]
I should add some tests.

File: AK/StringUtils.h (hunk 1 above)
Comment (howar6hill) [reply AK/StringUtils.h]: How about 
```C++
using AK::CaseSensitivity;
namespace StringUtils = AK::StringUtils;
```

Comment (awesomekling) [reply AK/StringUtils.h]: That would work, sure, but does anything outside of `AK` need to use `StringUtils` anyway? It's just meant to share code between `AK::String` and `AK::StringView`, right?

Comment (howar6hill) [reply AK/StringUtils.h]: So we don't have to type `AK`,  just as other classes in the `AK` namespace.

Comment (awesomekling) [reply AK/StringUtils.h]: You don't need to type `AK::` if you are within `namespace AK { }` :)

**Comment by howar6hill** [comment]
Shared functionalities in the StringView and String classes can then be put in  `AK::StringUtils`.

File: AK/StringUtils.h (hunk 1 above)
Comment (howar6hill) [reply AK/StringUtils.h]: `matches()` could be called outside of the `AK` namespace, just as `String::to_int()`.

Comment (awesomekling) [reply AK/StringUtils.h]: Right, but `String::to_int()` is a member function of `String` whereas `matches()` is a free function with a very generic name. I'm saying let's not put such a generic name in the global namespace. :)

Comment (howar6hill) [reply AK/StringUtils.h]: It depends on how we intend to use  the `AK::StringUtils` namespace.
Should we use `AK::StringUtils` just for internal use?
If not, the name of `matches` could be changed to something more specific.

**Comment by howar6hill** [comment]
Made changes accordingly. Added some tests.

**Comment by howar6hill** [comment]
@awesomekling Hope this is OK.

**Comment by awesomekling** [comment]
LGTM!

**Pull request merged**